src/react/webgl-renderer.js

```diff
@@ -131,31 +126,53 @@ export default class WebGLRenderer extends React.Component {
     }
   }
 
-  // Updates WebGL viewport to latest props
-  // for clean logging, only calls gl.viewport if props have changed
-  _updateGLViewport() {
-    let {viewport: {x, y, width: w, height: h}} = this.props;
-    const {pixelRatio: dpr} = this.props;
-    const {gl} = this;
-
-    x = x * dpr;
-    y = y * dpr;
-    w = w * dpr;
-    h = h * dpr;
+  // Calculate the drawing buffer size that would cover current canvas size and device pixel ratio
+  // Intention is that every pixel in the drawing buffer will have a 1-to-1 mapping with
+  // actual device pixels in the hardware framebuffer, allowing us to render at the full
+  // resolution of the device.
+  _calculateDrawingBufferSize(canvas, {useDevicePixelRatio = true}) {
+    const cssToDevicePixels = useDevicePixelRatio ? window.devicePixelRatio || 1 : 1;
+
+    // Lookup the size the browser is displaying the canvas in CSS pixels
+    // and compute a size needed to make our drawingbuffer match it in
+    // device pixels.
+    return {
+      width: Math.floor(canvas.clientWidth * cssToDevicePixels),
+      height: Math.floor(canvas.clientHeight * cssToDevicePixels),
```

Comment:
Just use `this.props.width` and `this.props.height`?
`clientWidth` and `clientHeight` causes page reflow. We already know the canvas css size since we set it here in `render()`.